apollo-runtime/src/main/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransport.kt

```diff
@@ -0,0 +1,166 @@
+package com.apollographql.apollo.subscription
+
+import com.apollographql.apollo.api.internal.json.JsonWriter
+import com.apollographql.apollo.subscription.OperationClientMessage.Start.Companion.jsonValue
+import okhttp3.HttpUrl
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.WebSocket
+import okio.Buffer
+import java.lang.ref.WeakReference
+import java.util.Base64
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * [SubscriptionTransport] implementation based on [WebSocket].
+ */
+class WebSocketSubscriptionTransport @JvmOverloads constructor(
+    private val webSocketRequest: Request,
+    private val webSocketConnectionFactory: WebSocket.Factory,
+    private val callback: SubscriptionTransport.Callback,
+    private val writePayloadAsJsonString: Boolean = false,
+    private val extensions: Map<String, Any?> = emptyMap()
+) : SubscriptionTransport {
+  internal val webSocket = AtomicReference<WebSocket>()
+  internal val webSocketListener = AtomicReference<WebSocketListener>()
+
+  override fun connect() {
+    val webSocketListener = WebSocketListener(this)
+    check(this.webSocketListener.compareAndSet(null, webSocketListener)) {
+      "Already connected"
+    }
+    webSocket.set(webSocketConnectionFactory.newWebSocket(webSocketRequest, webSocketListener))
+  }
+
+  override fun disconnect(message: OperationClientMessage) {
+    webSocket.getAndSet(null)
+        ?.close(1001, message.serializeToJson())
+    release()
+  }
+
+  override fun send(message: OperationClientMessage) {
+    val socket = webSocket.get() ?: run {
+      callback.onFailure(IllegalStateException("Send attempted on closed connection"))
+      return
+    }
+    socket.send(message.serializeToJson())
+  }
+
+  internal fun onOpen() {
+    callback.onConnected()
+  }
+
+  internal fun onMessage(message: OperationServerMessage?) {
+    callback.onMessage(message)
+  }
+
+  internal fun onFailure(t: Throwable?) {
+    try {
+      callback.onFailure(t)
+    } finally {
+      release()
+    }
+  }
+
+  internal fun onClosed() {
+    try {
+      callback.onClosed()
+    } finally {
+      release()
+    }
+  }
+
+  private fun release() {
+    webSocketListener.getAndSet(null)?.release()
+    webSocket.set(null)
+  }
+
+  private fun OperationClientMessage.serializeToJson(): String =
+      toJsonString(
+        writePayloadAsJsonString = writePayloadAsJsonString,
+        extensions = extensions
+    )
+
+  internal class WebSocketListener(delegate: WebSocketSubscriptionTransport) : okhttp3.WebSocketListener() {
```

Comment:
I kept this as `internal` since I'm assuming it's there to avoid the synthetic accessor methods from being generated?

Reply:
I think this used to be package private in Java. This could certainly be `private` except we're using it from tests so `internal` seems good